ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -0,0 +1,350 @@
+    compressionConfig = config;
+    compressionMetrics = metrics;
+  }
+
+  /**
+   * Get the compression map that contains all compression algorithms.
+   * @return The compression map.
+   */
+  public CompressionMap getAllCompressions() {
+    return allCompressions;
+  }
+
+  /**
+   * Check whether the blob is compressible based on the blob properties.
+   *
+   * @param blobProperties The blob property that is common to all chunks. It helps to detect whether blob is
+   *                       compressible or not.
+   * @return True if blob is compressible based on the blob properties; FALSE if blob is not compressible.
+   */
+  public boolean isBlobCompressible(BlobProperties blobProperties) {
+    Objects.requireNonNull(blobProperties, "blobProperties cannot be null.");
+
+    // Return false if compression is disabled.  No need to emit metrics.
+    if (!compressionConfig.isCompressionEnabled) {
+      return false;
+    }
+
+    // Check content-encoding. BlobProperty.ContentEncoding contains a string, do not apply compression.
+    // Note: Calling getBlobProperties() throws if blob properties is null, so it never returns null.
+    String contentEncoding = blobProperties.getContentEncoding();
+    if (compressionConfig.isSkipWithContentEncoding && !Utils.isNullOrEmpty(contentEncoding)) {
+      logger.info("No compression applied because compression content is encoded with " + contentEncoding);
+      // Emit metrics to count how many compressions skipped due to encoded data.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipContentEncoding.inc();
+      return false;
+    }
+
+    // Check the content-type.
+    String contentType = blobProperties.getContentType();
+    if (!compressionConfig.isCompressibleContentType(contentType)) {
+      logger.info("No compression applied because content-type " + contentType + " is incompressible.");
+      // Emit metrics to count how many compressions skipped due to incompressible chunks based on content-type.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipContentTypeFiltering.inc();
+      return false;
+    }
+
+    // For all other case, compression should be enabled.
+    return true;
+  }
+
+  /**
+   * Apply compression to a chunk buffer if compression is enabled and content-type is compressible.
+   * It applies compression config to determine whether to accept or reject the compressed result.
+   * For example, if the compressed data size is smaller than original data by certain % (exceeds threshold),
+   * the compressed data is accepted and return compressed data; otherwise, compression is discarded.
+   * This method emits metrics along the process.  It returns null if failed instead of throwing exceptions.
+   * This method is typically called for the first chunk to determine whether to compress the entire blob or not.
```

Comment:
how do u extrapolate this from just the first data chunk ?

Reply:
Sorry, it was out-dated comment.  Let me remove it since it causes confusion.  Compression is applied per chunk basis, not per blob.